File: 2018/p11.c
// p11.c

// Written by Jakob Ruhe 2018-12-11

// Build with:
// gcc --std=c99 -O2 -Wall p11.c -o p11

#include <stdio.h>
#include <string.h>

#define GridSize 300

typedef int value_t;

value_t calc_power_level(int x, int y, value_t serial)
{
  value_t rack_id = x + 10;
  value_t power_level = rack_id * y;
  power_level += serial;
  power_level *= rack_id;
  value_t h = (power_level / 100) % 10 - 5;
  return h;
}

value_t square(int minx, int miny, int size, value_t serial)
{
  value_t sum = 0;
  for (int y = miny; y < miny + size; y++) {
    for (int x = minx; x < minx + size; x++) {
      sum += calc_power_level(x, y, serial);
    }
  }
  return sum;
}

void solve1(value_t serial)
{
  value_t best_value;
  int best_x = -1, best_y, best_z;
  int z = 3;
  for (int miny = 1; miny < GridSize - z + 1; miny++) {
    for (int minx = 1; minx < GridSize - z + 1; minx++) {
      value_t value = square(minx, miny, z, serial);
      if (best_x == -1 || (value > best_value)) {
        best_value = value;
        best_x = minx;
        best_y = miny;
        best_z = z;
      }
    }
  }
  printf("solve1(%d): %d,%d, power level: %d\n",
      serial, best_x, best_y, best_value);
}

void solve2(value_t serial)
{
  value_t best_value;
  int best_x = -1, best_y, best_z;
  for (int miny = 1; miny < GridSize + 1; miny++) {
    for (int minx = 1; minx < GridSize + 1; minx++) {
      int zmax = miny > minx ? GridSize - miny + 1: GridSize - minx + 1;
      value_t value = calc_power_level(minx, miny, serial);
      for (int z = 1; z <= zmax; z++) {
        if (z >= 2) {
          int x, y;
          y = miny + z - 1;
          for (x = minx; x < minx + z; x++) {
            value += calc_power_level(x, y, serial);
          }
          x = minx + z - 1;
          for (y = miny; y < miny + z; y++) {
            value += calc_power_level(x, y, serial);
          }
        }
        if (best_x == -1 || (value > best_value)) {
          best_value = value;
          best_x = minx;
          best_y = miny;
          best_z = z;
        }
      }
    }
  }
  printf("solve2(%d): %d,%d,%d, power level: %d\n",
      serial, best_x, best_y, best_z, best_value);
}

int main()
{
  solve1(18);
  solve1(42);
  solve1(7672);
  solve2(18);
  solve2(42);
  solve2(7672);
  return 0;
}

File: 2018/p11.py
#!/usr/bin/env python3

# By Jakob Ruhe 2018-12-11

import unittest

def calc_power_level(x, y, serial):
    rack_id = x + 10
    power_level = rack_id * y
    power_level += serial
    power_level *= rack_id
    h = (power_level // 100) % 10
    value = h - 5
    return value

def square(minx, miny, size, serial):
    s = 0
    for y in range(miny, miny + size):
        for x in range(minx, minx + size):
            s += calc_power_level(x, y, serial)
    return s

def solve1(serial):
    best_value = None
    best_xy = None
    total_size = 300
    square_size = 3
    for miny in range(1, total_size - square_size + 1):
        for minx in range(1, total_size - square_size + 1):
            s = square(minx, miny, square_size, serial)
            if best_value is None or s > best_value:
                best_value = s
                best_xy = (minx, miny)
    return (best_xy, best_value)

class TestThis(unittest.TestCase):
    def test_power_level(self):
        self.assertEqual(calc_power_level(3, 5, 8), 4)
        self.assertEqual(calc_power_level(122, 79, 57), -5)
        self.assertEqual(calc_power_level(217, 196, 39), 0)
        self.assertEqual(calc_power_level(101, 153, 71), 4)
    def test_solve1(self):
        self.assertEqual(solve1(18), ((33, 45), 29))
        self.assertEqual(solve1(42), ((21, 61), 30))

if __name__ == "__main__":
    #unittest.main()
    (p1, b1) = solve1(7672)
    print("The answer to subproblem 1 is: {} ({})".format(",".join(map(lambda e: str(e), p1)), b1))
    print("Use the solution written in C to solve subproblem 2.")
